packages/ContentAutotagging/Engine/generic/AutotagBaseEngine.ts

```diff
@@ -0,0 +1,745 @@
+            let prompt = ''
+            for (const result of results.Results) {
+                prompt += `${result.Get('Prompt')}\n`
+            }
+
+            return prompt
+        }
+        return ''
+    }
+    
+    /**
+    * Given the content source parameters, this function creates a description of the content source item.
+    * @param contentSourceParams: The parameters of the content source item
+    * @returns The description of the content source item
+    */
+    public async getContentItemDescription(contentSourceParams: ContentSourceParams, contextUser): Promise<string> {
+        const description = `${await this.getContentTypeName(contentSourceParams.ContentTypeID, contextUser)} in ${await this.getContentFileTypeName(contentSourceParams.ContentFileTypeID, contextUser)} format obtained from a ${await this.getContentSourceTypeName(contentSourceParams.ContentSourceTypeID, contextUser)} source`
+        
+        return description
+    }
+
+    public async getChecksumFromURL(url: string): Promise<string> {
+        const response = await axios.get(url)
+        const content = response.data
+        const hash = crypto.createHash('sha256').update(content).digest('hex')
+        return hash
+    }
+
+    public async getChecksumFromText(text: string): Promise<string> {
+        const hash = crypto.createHash('sha256').update(text).digest('hex')
+        return hash
+    }
+
+    public async getContentItemIDFromURL(contentSourceParams: ContentSourceParams, contextUser: UserInfo): Promise<string> {
+        const url = contentSourceParams.URL
+        const rv = new RunView()
+        try{
+            const results = await rv.RunView<ContentItemEntity>({
+                EntityName: 'Content Items',
+                ExtraFilter: `URL='${url}' AND ContentSourceID='${contentSourceParams.contentSourceID}'`,
+                ResultType: 'entity_object'
+            }, contextUser)
+
+            if (results.Success && results.Results.length) {
+                return results.Results[0].Get('ID')
+            }
+            else {
+                throw new Error(`Content item with URL ${url} not found`)
+            }
+        }
+        catch (e) {
+            throw new Error(`Failed to retrieve content item ID from URL: ${url}`)
+        }
+    }
+
+    /**
+     * Given the results of the processing from the LLM, this function saves the details of the process run in the database.
+     * @param processRunParams: The parameters holding the details of the process run
+     * @param contextUser: The user context to save the process run
+     * @returns
+     */
+    public async saveProcessRun(processRunParams: ProcessRunParams, contextUser: UserInfo){
+        const md = new Metadata()
+        const processRun = <ContentProcessRunEntity> await md.GetEntityObject('Content Process Runs', contextUser)
+        processRun.NewRecord()
+        processRun.Set('SourceID', processRunParams.sourceID)
```

Comment:
same comment on weak typing. You also should be doing

`await md.GetEntityObject<ContentProcessRunEntity>('Content Process Runs', contextUser)`

INSTEAD OF

`<ContentProcessRunEntity> await md.GetEntityObject('Content Process Runs', contextUser)`

Either way the return variable from that method is a strongly typed sub-class that has members like `SourceID` so you can say:

`processRun.SourceID = processRunParams.sourceID` which is way better code

@nico-oz go through all of your code and switch to strong typing before we release this today please. Very important for maintainability and more robust code.

Reply:
`strong typing` may be an overstatement in this case. Using explicit generic types `<ContentProcessRunEntity>` is still a type-cast– there's no safety at runtime that the object returned is that type at all. However, it _does_ help in the IDE. Just have to keep in mind that the IDE is giving you some (potentially) false assurances about the object you're working with.